File: java/src/main/java/org/urlshortener/dal/repositories/impl/UserRepositoryImpl.java
package org.urlshortener.dal.repositories.impl;

import org.urlshortener.dal.entities.User;
import org.urlshortener.dal.repositories.UserRepository;

import javax.enterprise.context.RequestScoped;
import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

@RequestScoped
public class UserRepositoryImpl implements UserRepository {

    @PersistenceContext
    private final EntityManager em;

    public UserRepositoryImpl(EntityManager em) {
        this.em = em;
    }

    @Override
    public User merge(User user)  {
        return this.em.merge(user);
    }

    @Override
    public void deleteById(Long id) {
        var query = this.em.createQuery("delete from User u where u.id = :id");
        query.setParameter("id", id);
        query.executeUpdate();
    }

    @Override
    public User findByName(String name) {
        var query = this.em
                .createQuery("select u from User u where u.name = :name", User.class);
        query.setParameter("name", name);
        return query.getSingleResult();
    }

    @Override
    public User findById(Long userId) { return this.em.find(User.class, userId); };

    @Override
    public boolean existsByName(String name) {
        var query = this.em
                .createQuery("select count(u) from User u where u.name = :name", Long.class);
        query.setParameter("name", name);
        return query.getSingleResult() != 0;
    }
}

File: src/main/kotlin/com/urlshortener/dal/repositories/impl/UserRepositoryImpl.kt
package com.urlshortener.dal.repositories.impl

import com.urlshortener.dal.entities.User
import com.urlshortener.dal.repositories.UserRepository
import javax.enterprise.context.RequestScoped
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext

@RequestScoped
class UserRepositoryImpl(@PersistenceContext private val em: EntityManager) : UserRepository{
    override fun merge(user: User): User = em.merge(user)

    override fun deleteById(id: Long) {
        em.createQuery("delete from User u where u.id = :id").run {
            setParameter("id", id)
            executeUpdate()
        }
    }

    override fun findByName(name: String): User? =
        em.createQuery("select u from User u where u.name = :name", User::class.java).run {
            setParameter("name", name)
            singleResult
        }

    override fun findById(userId: Long): User? = em.find(User::class.java, userId)

    override fun existsByName(name: String): Boolean =
        em.createQuery("select count(u) from User u where u.name = :name", java.lang.Long::class.java)
            .run {
                setParameter("name", name)
                singleResult.toLong() != 0L
            }
}